Fixed "Email Copy of Invoice", "Email Copy of Shipment" and "Email Copy of Creditmemo" Issues

gwharton opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Magento is not consistent with the operation of the "Email Copy of ......" Checkboxes on the Invoice, Creditmemo and Shipment creation screens in admin.

On the invoice screen, if the box is unticked, but the global setting in Sales Emails is set to enabled, the email still gets sent.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#28511

### Manual testing scenarios (*)

- Create test Product
- Purchase Test Product
- Set Config -> Sales -> Sales Emails to Enabled for Invoice, Shipment and Creditmemo
- Create Invoice and do NOT tick "Email copy of Invoice". Verify Email is NOT received
- Create Shipment and do NOT tick "Email copy of Shipment". Verify Email is NOT received
- Create Creditmemo and do NOT tick "Email copy of Creditmemo". Verify Email is NOT received
- Purchase Test Product
- Create Invoice and tick "Email copy of Invoice". Verify Email is received
- Create Shipment and tick "Email copy of Shipment". Verify Email is received
- Create Creditmemo and tick "Email copy of Creditmemo". Verify Email is received
- Purchase Test Product
- Set Config -> Sales -> Sales Emails to Disabled for Invoice, Shipment and Creditmemo
- Create Invoice. Verify Email is NOT received
- Create Shipment. Verify Email is NOT received
- Create Creditmemo. Verify Email is NOT received

### Questions or comments
Unit test covered

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

gwharton commented:
@magento run all tests

gwharton commented:
Functional Tests EE

gwharton commented:
@magento run Functional Tests EE

gwharton commented:
@magento run Functional Tests CE

gwharton commented:
@magento run Magento Health Index

gwharton commented:
@magento run all tests

gwharton commented:
@magento run all tests

ihor-sviziev commented:
Hi @gwharton,
Looks like we need to change the default value of the checkbox on the "send emdail... " to be according to the value in the admin. What do you thing?

gwharton commented:
The checkbox is only present if the global setting is enabled (which actually makes the additional check to see if it is enabled before sending email a little redundant), however......

So the desired operation is?

If the global setting is enabled, then the checkbox is visible, and checked by default. The user can uncheck if he doesn't want to send the email? If the global setting is disabled, the checkbox is not present and the email is not sent.

That sounds fair to me.

gwharton commented:
@magento run all tests

gwharton commented:
@magento run Functional Tests CE

gwharton commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
@gwharton,
I believe the decision what is correct behavior isn't really simple.

Technically "can send invoice email" or "can send credit memo email" should be kind of global. Seems like this global option was designed for automated email sending while such entity was created, but the checkbox in admin - for manual.
So in theory earlier we were able to not send such emails automatically, but still were able to send an email if we would like to notify our customer.

I though doing smart thing - if the option is enabled - let's check the checkbox, otherwise - do not check it by default, but seems like I was wrong.

Unfortunately I don't have enough time now in order to collect all the possible use cases in this area, but I believe all of them need to be checked to make sure we're not breaking any existing flow.

gwharton commented:
No problem. Happy to standby. Its pretty easy to implement the logic. Its coming up with the right logic thats the difficult bit. Will await further instruction.

aleron75 commented:
@magento give me 2.4-develop instance

gwharton commented:
I'm just going to drop a brain dump for you all to promote conversation on the item.

Currently the system works as follows.

```
- Global Setting for Invoice Emails is disabled
-- Create Invoice
--- Checkbox to send email is not visible. 
---- Save Invoice. No email is sent on invoice creation.

- Global Setting for Invoice Emails is enabled
-- Create Invoice
--- Checkbox to send email is unchecked (default).
---- Save Invoice. Email is sent despite checkbox being unchecked <--- NOT RIGHT.
-- Create Invoice
--- Checkbox to send email is checked.
---- Save Invoice. Email is sent.
```

```
- Global Setting for CreditMemo Emails is disabled
-- Create CreditMemo 
--- Checkbox to send email is not visible. 
---- Save CreditMemo. No email is sent on CreditMemo creation.

- Global Setting for CreditMemo Emails is enabled
-- Create CreditMemo 
--- Checkbox to send email is unchecked (default).
---- Save CreditMemo. Email is not sent.
-- Create CreditMemo 
--- Checkbox to send email is checked.
---- Save CreditMemo. Email is sent.
```

```
- Global Setting for Shipment Emails is disabled
-- Create Shipment 
--- Checkbox to send email is not visible. 
---- Save Shipment. No email is sent on Shipment creation.

- Global Setting for Shipment Emails is enabled
-- Create Shipment 
--- Checkbox to send email is unchecked (default).
---- Save Shipment. Email is not sent.
-- Create Shipment 
--- Checkbox to send email is checked.
---- Save Shipment. Email is sent.
```


The PR proposes two changes. 

- The default for the checkbox on the create invoice/creditmemo/shipment page would change to be checked by default. You can still untick it to not send the email. If the global setting is disabled, the checkbox doesn't even appear.
- The inconsistency with the invoice page would be corrected, bringing it inline with how the shipment and creditmemo pages work. Currently the invoice page shows the checkbox, but actually ignores the contents of the checkbox so always sends the invoice (if the global setting is enabled). This PR would remove that ability. Was this put in on purpose?

So the question that now arrises is ....

"Do we want to be able to send emails manually from the invoice/creditmemo/shipment creation pages when the global setting for emails (for the specified type) is disabled?
"Should the global setting effectively disable the sending of those email types globally (currently doesn't)"
"Do we want to decouple the decision of whether to send the email when creating the invoice/creditmemo/shipment (i.e the checkbox on the creation page) from the global setting?"

i.e 

- Global setting is disabled, checkbox visible on create page, and if ticked, the email is sent (currently can't do this).
- Global setting is enabled, checkbox visible on create page, and if unticked, the email is not sent (currently can't do this on invoice)

I like the idea of decoupling the global enabled setting from the creation page checkbox, so when you are creating the invoice/creditmemo/shipment it just does what the checkbox says. If this is the case, then further work needs doing on the config page

- Renaming the setting as it is no longer a global enable/disable flag.
- The other settings like template, and copy to, need to be decoupled from the enabled setting as these would still need to be set, even though the global setting is disabled.
- Other areas may need looking at where the global setting is used.
- Does the "Send Email" button at the top of the sales order view pages work properly. Does that always override the global setting?

sidolov commented on app/code/Magento/Sales/view/adminhtml/templates/order/creditmemo/create/items.phtml:
```diff
@@ -126,7 +126,8 @@ $commentText = $block->getCreditmemo()->getCommentText();
                            value="1"
-                           type="checkbox" />
+                           type="checkbox"
+                           checked />
```
Could you please explain why it's checked by default? It's definitely backward incompatible change since the default behavior is changed

gwharton replied:
Sure, It was something that @ihor-sviziev suggested, however since then, with more thought, it is not so clear whether this is the right behaviour. Like you say, it changes the default behaviour.

At the moment, the checkbox is only visible if the global setting to enable emails is set. So, if it is not set globally, the checkbox is hidden, and the email doesn't get sent. If the global setting is enabled, then the checkbox is visible and the user can choose whether to send the email or not.

More than happy to revert that change back though and leave the default to unchecked, so the admin has to have global setting enabled, and also tick the box on every invoice/creditmemo/shipment to send the email. If the global setting is enabled, and the box is left unchecked, then the email doesn't get sent. This seems the least contentious solution.

ihor-sviziev replied:
Hi @gwharton ,
I just analyzed all use cases (thx for good and detailed explanation) - seems like we should not check the checkbox by default. Such approach will be more clear and backward compatible. 
Could you revert this change? Sorry for confusing you last time

ihor-sviziev reviewed (changes requested):
Please revert changes in setting checked state for checkboxes

gwharton commented:
@magento run all tests

gwharton commented:
@magento run Functional Tests B2B

gwharton commented:
@magento run Integration Tests

gwharton commented:
@magento run Functional Tests B2B

ihor-sviziev commented on app/code/Magento/Sales/Test/Unit/Controller/Adminhtml/Order/Invoice/SaveTest.php:
```diff
@@ -98,11 +130,38 @@ protected function setUp(): void
             Save::class,
             [
                 'context' => $contextMock,
+                'registry' => $registryMock,
```
please just remove it, it's not actually used and causing MHI test failure that will require additional approval.

gwharton commented:
@magento run all tests

gwharton commented:
Let me know if you would like me to squash commit history.

ihor-sviziev commented:
Yeah, would be good

On Tue, 17 Nov 2020 at 19:41 Graham Wharton <notifications@github.com>
wrote:

> Let me know if you would like me to squash commit history.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/magento2/pull/30868#issuecomment-729089384>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOJOULHNZXWXPZHZJUSGATSQKYT5ANCNFSM4TREYDHQ>
> .
>

Commit:
Updated method used to determine whether emails are sent on creation of invoices,
creditmemos and shipments from the admin page. Now takes into account global "enabled"
setting and the checkbox on creation page.

Updated Unit and Integration tests to suit.

gwharton commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests CE

ihor-sviziev commented:
I just wanted to wait till the test results, everything looks good to me

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-8472](https://jira.corp.magento.com/browse/ENGCOM-8472) has been created to process this Pull Request

ihor-sviziev reviewed (approved):
✔ Approved.

Failing tests looks not related to changes form this PR.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8472](https://jira.corp.magento.com/browse/ENGCOM-8472) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

### Main Case

**Manual testing scenario:**

1. Go to Admin -> Stores -> Configuration -> Sales -> Sales Emails -> Invoice and
set Enable to Yes;
2. Create Order;
3. Go to Admin -> Sales -> Orders and open created before order for edit;
4. Click on the Invoice button;
5. Uncheck "Email Copy of Invoice";
6. Generate Invoice;

**Before:** :heavy_multiplication_x:  Email sent to Customer

**After:** :heavy_check_mark: Email was not sent to Customer

### Additional Cases

- _**Manual testing scenario:**_

1. Create test Product
2. Purchase Test Product
3. Set Config -> Sales -> Sales Emails to Enabled for Invoice, Shipment and Creditmemo
Create Invoice and do NOT tick "Email copy of Invoice". 

**Actual Result**: :heavy_check_mark:  An email was NOT received

4. Create Shipment and do NOT tick "Email copy of Shipment". 

**Actual Result**: :heavy_check_mark:  An email was NOT received

5. Create Creditmemo and do NOT tick "Email copy of Creditmemo". 

**Actual Result**: :heavy_check_mark:  An email was NOT received


- _**Manual testing scenario:**_

1. Purchase Test Product
2. Create Invoice and tick "Email copy of Invoice".

**Actual Result**: :heavy_check_mark:  An email was received

3. Create Shipment and tick "Email copy of Shipment".

**Actual Result**: :heavy_check_mark:  An email was received

4. Create Creditmemo and tick "Email copy of Creditmemo". 

**Actual Result**: :heavy_check_mark:  An email was received

- _**Manual testing scenario:**_

1. Purchase Test Product
2. Set Config -> Sales -> Sales Emails to Disabled for Invoice, Shipment and Creditmemo
3. Create Invoice. Verify Email is NOT received

**Actual Result**: :heavy_check_mark:  An email was NOT received

4. Create Shipment. Verify Email is NOT received

**Actual Result**: :heavy_check_mark:  An email was NOT received

5. Create Creditmemo. Verify Email is NOT received

**Actual Result**: :heavy_check_mark:  An email was NOT received

engcom-Alfa commented:
:heavy_check_mark: QA Passed
